torchgen/dest/register_dispatch_key.py

```diff
@@ -409,6 +416,77 @@ def generate_defn(cpp_sig: CppSignature) -> str:
                 metadata = self.backend_index.get_kernel(f)
                 if metadata is None:
                     return None
+
+                # define the struct for dispatch-less composite kernels.
+                impl_struct = ""
+                if self.backend_index.should_gen_dispatchless_composite(f):
+                    # define the native function as a wrapper static method, redirecting the call to
+                    # the appropriate function (e.g. on CPU, on CUDA, or as a CompositeExplicitAutograd).
+                    def dispatchless_function_defn(
+                        f: NativeFunction, g: Optional[NativeFunctionsGroup]
+                    ) -> str:
+                        dispatch_key = self.backend_index.dispatch_key
+
+                        # the declaration signature is a static method (not a Tensor method).
+                        decl_sig = CppSignatureGroup.from_native_function(
+                            f, method=False, fallback_binding=False
+                        ).signature
+
+                        # decide 2 things:
+                        #     1. where is the redirect function? (i.e. its namespace)
+                        #     2. is it a Tensor method? (some operations are only available as methods)
+                        if self.backend_index.has_registered_kernel(f, g):
+                            prefix = f"at::{dispatch_key.lower()}::"
+                            is_method = False
+                        else:
+                            # give preference to function variants instead of method.
+                            is_method = {Variant.method} == f.variants
+                            if Variant.function in f.variants:
+                                if (
+                                    dispatch_key
+                                    == DispatchKey.CompositeExplicitAutograd
+                                ):
+                                    prefix = "at::"
+                                else:
+                                    # dependent dispatch-less composite kernel headers are not included
+                                    # for CPU or CUDA. However, we shouldn't hit this point, since they
+                                    # should have a dispatch-less kernel generated at this dispatch key.
+                                    prefix = "at::native::"
+                            else:
+                                assert f.func.arguments.self_arg is not None
+                                prefix = f"{f.func.arguments.self_arg.argument.name}."
+
+                        # the call-site signature depends on whether we are calling it as
+                        # a method or not.
+                        call_sig = CppSignatureGroup.from_native_function(
+                            f, method=is_method, fallback_binding=False
+                        ).signature
```

Comment:
You'll also probably want to get `.most_faithful_signature()` here instead of just `.signature`, since the faithful signature will line up more closely with the native API (e.g. you won't need to re-wrap TensorOptions arguments, or move out= arguments around, when calling into `at::cpu::{op}`)

Reply:
Hmm... I think we'd better keep using just `.signature`, here. The reason being that whenever a kernel writers specify a dependency on, say, `neg_out`, I'm assuming they are using the `neg_out` C++ API call, i.e. the non-faithful one. What do you think?